### PR TITLE
Feat/transpile controlled parametric gates

### LINKED ISSUE #358: Fix: Preserve global phase
# Overview

Snowflurry currently ignores the global phase, but the global phase matters when performing control gates.

## Tasks

- [ ] Ensure that the global phase is preserved for all operators
- [ ] Verify that `controlled(phase_shift(2,pi/2),[1])` and `controlled(z_90(2),[1])` are not equivalent.

### PR DESCRIPTION
**Pull request by fSosaRey**
closes #358

### CODE REVIEW

File: src/core/transpile.jl [hunk 1]
```diff
@@ -914,6 +914,41 @@ function cast_to_rz_rx_rz(gate::Universal, target::Int)::Vector{Gate}
...
+    γ = params["theta"]
+    β = params["phi"]
+    δ = params["lambda"]
+
+    A = rotation_z(β) * rotation_y(γ / 2.0)
+
+    B = rotation_y(-γ / 2.0) * rotation_z(-(δ + β) / 2.0)
+
+    # numerical stability issues arise for rotation_z with vanishingly small phase angle
+    if abs(δ - β) > sqrt(eps(typeof(δ)))
```
Comment (plerouxanyon) [inline src/core/transpile.jl:934]: I don't undestand how this gives numerical instability issues.

Given an approximate Taylor series expansion about zero of `exp(-im * x)` (The exponent of the `rotation_z`)

image

I get identical results

image

Could you add a comment on why this gives numerical instability issues, as it's not clear from these 5 lines to me?

File: src/core/transpile.jl [hunk 2]
```diff
@@ -2064,3 +2101,68 @@ function transpile(
...
+            connected_qubits = get_connected_qubits(instr)
+            if length(connected_qubits) != 2
+                throw(
+                    ArgumentError(
+                        "DecomposeControlledGatesTranspiler is only implemented for single-target single-control Controlled Gates",
+                    ),
+                )
+            end
+
+            if typeof(kernel) == SigmaZ
```
Comment (plerouxanyon) [inline src/core/transpile.jl:2133]: Perhaps we should just push a `control_z` gate here and let a later stage transpile that.

File: src/core/transpile.jl [hunk 3]
```diff
@@ -2064,3 +2101,68 @@ function transpile(
...
+                throw(ArgumentError("use control_z() instead of Controlled(SigmaZ)"))
+            end
+
+            if typeof(kernel) == SigmaX
```
Comment (plerouxanyon) [inline src/core/transpile.jl:2137]: Same as above, but with `control_x`

File: src/core/transpile.jl [hunk 4]
```diff
@@ -2064,3 +2101,68 @@ function transpile(
...
+            # store existing global phase, which is cancelled by as_universal_gate()
+            a_0 = get_canonical_global_phase(get_matrix(op))
+            universal_equivalent = as_universal_gate(target, op)
+
+            (a_d, A, B, C) = decompose_universal_to_A_B_C_gates(
+                get_gate_symbol(universal_equivalent),
+                target,
+            )
+
+            α = -a_0 + a_d
```
Comment (plerouxanyon) [inline src/core/transpile.jl:2155]: Let's use `\alpha` or `a` in both places for consistency.

File: src/core/transpile.jl [hunk 5]
```diff
@@ -2064,3 +2101,68 @@ function transpile(
 
     return circuit
 end
+
+struct DecomposeControlledGatesTranspiler <: Transpiler end
```
Comment (plerouxanyon) [inline src/core/transpile.jl:2105]: Perhaps we can rename this to `DecomposeSingleTargetSingleControlGatesTranspiler`, and then let everything pass through that is not that. Then, the `UnsupportedGatesTranspiler` can catch anything that wasn't transpiled. This way we are not as dependent on the ordering of transpilers. What do you think?

**Review by plerouxanyon — CHANGES_REQUESTED** [review]

File: src/core/transpile.jl (hunk 5 above)
Comment (fSosaRey) [reply src/core/transpile.jl:2105]: yes thats is better, updated

File: src/core/transpile.jl (hunk 2 above)
Comment (fSosaRey) [reply src/core/transpile.jl:2133]: updated

File: src/core/transpile.jl (hunk 3 above)
Comment (fSosaRey) [reply src/core/transpile.jl:2137]: updated

File: src/core/transpile.jl (hunk 4 above)
Comment (fSosaRey) [reply src/core/transpile.jl:2155]: updated

File: src/core/transpile.jl (hunk 1 above)
Comment (fSosaRey) [reply src/core/transpile.jl]: I took it out, and the error i was seeing is gone

**Pull request merged**